app/histories.go

```diff
@@ -30,23 +31,51 @@ func HistoriesHandler(app *App) func(c echo.Context) error {
 			return c.String(echo.ErrUnauthorized.Code, echo.ErrUnauthorized.Message)
 		}
 
-		// retrieve messages
 		messages := make([]*models.Message, 0)
 		if app.Defaults.MongoEnabled {
 			collection := app.Defaults.MongoMessagesCollection
-
+			var wg sync.WaitGroup
+			var mu sync.Mutex
+			// guarantees ordering in responses payload
+			topicsMessagesMap := make(map[string][]*models.MessageV2, len(authorizedTopics))
+			for _, topic := range authorizedTopics {
+				wg.Add(1)
+				go func(topicsMessagesMap map[string][]*models.MessageV2, topic string) {
+					topicMessages := mongoclient.GetMessagesV2(
+						c,
+						mongoclient.QueryParameters{
+							Topic:      topic,
+							From:       from,
+							Limit:      limit,
+							Collection: collection,
+						},
+					)
+					mu.Lock()
+					topicsMessagesMap[topic] = topicMessages
+					mu.Unlock()
+					wg.Done()
+				}(topicsMessagesMap, topic)
+			}
```

Comment:
Once this code is duplicated in app/histories_v2.go, what do you think of extracting this logic to an external func?

Reply:
Unnecessary extraction since "/histories" will be deprecated soon enough.